### ISSUE TITLE
Asynchronous DNS lookups

### ISSUE THREAD

**Issue by martinboers** [eclipse-paho/paho.mqtt.c#915]
I am using the Paho C++ client in an embedded application which has strict timing requirements - basically, the application cannot block for more than (say) 100ms before it causes problems. So, I thought the asynchronous Paho client would be ideal, because I "[want no calls to block at all](https://modelbasedtesting.co.uk/2013/10/13/which-paho-mqtt-c-api-to-use-and-some-history/)".

During testing we have found that there are certain circumstances, related to network disconnections, when some calls to the Paho asynchronous client block for a long period - in the example below, exactly 20 or 40 seconds.

**To Reproduce**

* In an application that uses the Paho C++ asyncronous client, establish a connection to a remote MQTT server with default connection parameters (in my case, I used the `async_publish` sample as a basis).
* Call the `is_connected` method every 100 ms.
* Disconnect the WAN cable from the client device.
* Wait for 5 minutes.

After the WAN cable is disconnected, it takes 2-3 minutes for `is_connected` to return `false` (that's fine). However, after that, the call to `is_connected` blocks for almost precisely 20 or 40 seconds, presumably during reconnection attempts.  After returning, `is_connected` calls do not block for the next few seconds, but then it blocks again. This cycle repeats until the WAN cable is reconnected and the connection to the server is automatically re-established.

**Expected behavior**

* The `is_connected` method always returns either `true` or `false`, but does not block.

**Trace**

In the following sections of the trace, the `is_connected` method blocks between 11:46:44 and 11:47:04.

```text
=========================================================
                   Trace Output
Product name: Eclipse Paho Asynchronous MQTT C Client Library
Version: 1.3.1
Build level: 2020-04-08T19:53:33Z
OpenSSL version: OpenSSL 1.1.1d  10 Sep 2019
OpenSSL flags: compiler: arm-pxc-linux-gnueabi-gcc  -mthumb -mfpu=neon -mfloat-abi=hard -mcpu=cortex-a9 -fstack-protector-strong  -D_FORTIFY_SOURCE=2 -Wformat -Wformat-security -Werror=format-security --sysroot=recipe-sysroot -O2 -pipe -g -feliminate-unused-debug-types -fmacro-prefix-map=                      -fdebug-prefix-map=                      -fdebug-prefix-map=                      -fdebug-prefix-map= -DOPENSSL_USE_NODELETE -DOPENSSL_PIC -DOPENSSL_CPUID_OBJ -DOPENSSL_BN_ASM_MONT -DOPENSSL_BN_ASM_GF
OpenSSL build timestamp: built on: Wed Apr  8 19:47:42 2020 UTC
OpenSSL platform: platform: linux-armv4
OpenSSL directory: OPENSSLDIR: "/usr/lib/ssl-1.1"
/proc/version: Linux version 4.14.115-rt59-pxc (oe-user@oe-host) (gcc version 8.3.0 (GCC)) #1 SMP PREEMPT RT Wed Apr 8 20:24:03 UTC 2020

=========================================================
19700101 000000.000 (3069164720)  (1)> Socket_outInitialize:131
19700101 000000.000 (3069164720)   (2)> SocketBuffer_initialize:94
19700101 000000.000 (3069164720)   (2)< SocketBuffer_initialize:98
19700101 000000.000 (3069164720)  (1)< Socket_outInitialize:144
19700101 000000.000 (3069164720)  (1)> SSLSocket_initialize:446
19700101 000000.000 (3069164720)   (2)> SSL_create_mutex:346
19700101 000000.000 (3069164720)   (2)< SSL_create_mutex:352 (0)
19700101 000000.000 (3069164720)   (2)> SSL_create_mutex:346
19700101 000000.000 (3069164720)   (2)< SSL_create_mutex:352 (0)
19700101 000000.000 (3069164720)  (1)< SSLSocket_initialize:492 (1)
19700101 000000.000 (3069164720)  (1)> MQTTStrncpy:851
19700101 000000.000 (3069164720)  (1)< MQTTStrncpy:861
19700101 000000.000 (3069164720)  (1)> MQTTStrncpy:851
19700101 000000.000 (3069164720)  (1)< MQTTStrncpy:861
19700101 000000.000 (3069164720)  (1)> MQTTPersistence_create:52
19700101 000000.000 (3069164720)  (1)< MQTTPersistence_create:98 (0)
19700101 000000.000 (3069164720)  (1)> MQTTPersistence_initialize:113
19700101 000000.000 (3069164720)  (1)< MQTTPersistence_initialize:121 (0)
19700101 000000.000 (3069164720)  (1)> MQTTAsync_restoreCommands:1002
20200701 114421.346 0 commands restored for client async_publish
20200701 114421.346 (3069164720)  (1)< MQTTAsync_restoreCommands:1040 (0)
20200701 114421.346 (3069164720)  (1)> MQTTPersistence_restoreMessageQueue:719
20200701 114421.346 0 queued messages restored for client async_publish
20200701 114421.346 (3069164720)  (1)< MQTTPersistence_restoreMessageQueue:758 (0)
20200701 114421.346 (3069164720) (0)< MQTTAsync_createWithOptions:654 (0)
20200701 114421.346 (3069164720) (0)> MQTTAsync_setConnected:2614
20200701 114421.346 (3069164720) (0)< MQTTAsync_setConnected:2626 (0)
20200701 114421.346 (3069164720) (0)> MQTTAsync_setCallbacks:2500
20200701 114421.346 (3069164720) (0)< MQTTAsync_setCallbacks:2514 (0)
20200701 114421.346 (3069164720) (0)> MQTTAsync_connect:2812
20200701 114421.346 (3069164720)  (1)> Thread_start:69
20200701 114421.346 (3069164720)  (1)< Thread_start:79
20200701 114421.346 (3069164720)  (1)> Thread_start:69
20200701 114421.346 (3062973520) (0)> MQTTAsync_sendThread:1858
20200701 114421.346 (3069164720)  (1)< Thread_start:79
20200701 114421.346 (3062973520)  (1)> Thread_wait_cond:413
20200701 114421.346 (3069164720)  (1)> MQTTStrncpy:851
20200701 114421.346 (3069164720)  (1)< MQTTStrncpy:861
20200701 114421.346 (3069164720)  (1)> MQTTAsync_addCommand:1050
20200701 114421.348 (3069164720)   (2)> Thread_signal_cond:395
20200701 114421.348 (3069164720)   (2)< Thread_signal_cond:400 (0)
20200701 114421.348 (3069164720)  (1)< MQTTAsync_addCommand:1084 (0)
20200701 114421.348 (3069164720) (0)< MQTTAsync_connect:3119 (0)
20200701 114421.348 (3069164720) (0)> MQTTProperties_free:366
20200701 114421.348 (3069164720) (0)< MQTTProperties_free:389
20200701 114421.348 (3069164720) (0)> MQTTProperties_free:366
20200701 114421.348 (3069164720) (0)< MQTTProperties_free:389
20200701 114421.348 (3062973520)  (1)< Thread_wait_cond:421 (0)
20200701 114421.348 (3062973520)  (1)> MQTTAsync_checkTimeouts:1804
20200701 114421.348 (3054580816) (0)> MQTTAsync_receiveThread:2138
20200701 114421.348 (3054580816)  (1)> MQTTAsync_cycle:3742
20200701 114421.348 (3054580816)   (2)> Socket_getReadySocket:237
20200701 114421.348 (3054580816)   (2)< Socket_getReadySocket:309 (0)
20200701 114421.348 (3054580816)   (2)> MQTTAsync_sleep:440
20200701 114421.348 (3062973520)  (1)< MQTTAsync_checkTimeouts:1852
20200701 114421.348 (3062973520)  (1)> MQTTAsync_processCommand:1398
20200701 114421.348 Connecting to serverURI test.mosquitto.org:1883 with MQTT version 4
20200701 114421.348 (3062973520)   (2)> MQTTProtocol_connect:114
20200701 114421.348 (3062973520)    (3)> MQTTProtocol_addressPort:58
20200701 114421.351 (3062973520)    (3)< MQTTProtocol_addressPort:90
20200701 114421.351 (3062973520)    (3)> Socket_new:668
20200701 114421.351 New socket 4 for test.mosquitto.org:1883, port 1883
20200701 114421.351 (3062973520)     (4)> Socket_addSocket:173
20200701 114421.351 (3062973520)      (5)> Socket_setnonblocking:85
20200701 114421.351 (3062973520)      (5)< Socket_setnonblocking:90 (0)
20200701 114421.351 (3062973520)     (4)< Socket_addSocket:196 (0)
20200701 114421.351 Connect pending
20200701 114421.351 (3062973520)    (3)< Socket_new:785 (115)
20200701 114421.351 (3062973520)   (2)< MQTTProtocol_connect:156 (115)
20200701 114421.351 (3062973520)   (2)> MQTTProperties_free:366
20200701 114421.351 (3062973520)   (2)< MQTTProperties_free:389
20200701 114421.351 (3062973520)  (1)< MQTTAsync_processCommand:1734 (1)
20200701 114421.351 (3062973520)  (1)> Thread_wait_cond:413
20200701 114421.351 (3069164720) (0)> MQTTAsync_isConnected:3201
20200701 114421.351 (3069164720) (0)< MQTTAsync_isConnected:3206 (0)
20200701 114421.351 (3054580816)   (2)< MQTTAsync_sleep:446
20200701 114421.351 (3054580816)   (2)> MQTTAsync_retry:3575
20200701 114421.351 (3054580816)    (3)> MQTTProtocol_keepalive:608
20200701 114421.351 (3054580816)    (3)< MQTTProtocol_keepalive:644
20200701 114421.451 (3054580816)    (3)> MQTTProtocol_retry:730
20200701 114421.451 (3054580816)    (3)< MQTTProtocol_retry:749
20200701 114421.451 (3054580816)   (2)< MQTTAsync_retry:3585
20200701 114421.451 (3054580816)  (1)< MQTTAsync_cycle:3893 (-1)
20200701 114421.451 (3054580816)  (1)> MQTTAsync_cycle:3742
20200701 114421.451 (3054580816)   (2)> Socket_getReadySocket:237
20200701 114421.451 (3054580816)    (3)> Socket_continueWrites:928
20200701 114421.451 (3054580816)    (3)< Socket_continueWrites:952 (0)
20200701 114421.451 (3054580816)    (3)> isReady:213
20200701 114421.451 (3054580816)    (3)< isReady:218 (1)
20200701 114421.451 (3054580816)   (2)< Socket_getReadySocket:309 (4)
20200701 114421.451 m->c->connect_state = 1
20200701 114421.451 (3054580816)   (2)> MQTTAsync_connecting:3593
20200701 114421.451 (3054580816)    (3)> MQTTPacket_send_connect:55
20200701 114421.451 (3054580816)     (4)> MQTTPacket_send:187
20200701 114421.451 (3054580816)      (5)> MQTTPacket_encode:281
20200701 114421.451 (3054580816)      (5)< MQTTPacket_encode:291 (1)
20200701 114421.451 (3054580816)      (5)> WebSocket_putdatas:782
20200701 114421.451 (3054580816)       (6)> Socket_putdatas:502
20200701 114421.451 (3054580816)        (7)> Socket_writev:426
20200701 114421.453 (3054580816)        (7)< Socket_writev:478 (68)
20200701 114421.453 (3054580816)       (6)< Socket_putdatas:550 (0)
20200701 114421.453 (3054580816)      (5)< WebSocket_putdatas:814 (0)
20200701 114421.453 (3054580816)     (4)< MQTTPacket_send:216 (0)
20200701 114421.453 4 async_publish -> CONNECT version 4 clean: 1 (0)
20200701 114421.453 (3054580816)    (3)< MQTTPacket_send_connect:126 (0)
20200701 114421.453 (3054580816)   (2)< MQTTAsync_connecting:3732 (0)
20200701 114421.453 (3054580816)   (2)> MQTTAsync_retry:3575
20200701 114421.453 (3054580816)    (3)> MQTTProtocol_retry:730
20200701 114421.453 (3054580816)    (3)< MQTTProtocol_retry:749
20200701 114421.453 (3054580816)   (2)< MQTTAsync_retry:3585
20200701 114421.453 (3054580816)  (1)< MQTTAsync_cycle:3893 (0)
20200701 114421.453 (3054580816)  (1)> MQTTAsync_cycle:3742
20200701 114421.453 (3054580816)   (2)> Socket_getReadySocket:237
20200701 114421.453 (3054580816)    (3)> Socket_continueWrites:928
20200701 114421.453 (3054580816)    (3)< Socket_continueWrites:952 (0)
20200701 114421.453 (3054580816)    (3)> isReady:213
20200701 114421.453 (3054580816)    (3)< isReady:218 (1)
20200701 114421.453 (3054580816)   (2)< Socket_getReadySocket:309 (4)
20200701 114421.453 m->c->connect_state = 4
20200701 114421.483 (3054580816)   (2)> MQTTPacket_Factory:111
20200701 114421.483 (3054580816)    (3)> WebSocket_getch:476
20200701 114421.483 (3054580816)     (4)> Socket_getch:324
20200701 114421.483 (3054580816)      (5)> SocketBuffer_getQueuedChar:203
20200701 114421.483 (3054580816)      (5)< SocketBuffer_getQueuedChar:222 (-22)
20200701 114421.483 (3054580816)      (5)> SocketBuffer_queueChar:288
20200701 114421.483 (3054580816)      (5)< SocketBuffer_queueChar:313
20200701 114421.483 (3054580816)     (4)< Socket_getch:345 (0)
20200701 114421.483 (3054580816)    (3)< WebSocket_getch:513 (0)
20200701 114421.483 (3054580816)    (3)> MQTTPacket_decode:310
20200701 114421.483 (3054580816)     (4)> WebSocket_getch:476
20200701 114421.483 (3054580816)      (5)> Socket_getch:324
20200701 114421.483 (3054580816)       (6)> SocketBuffer_getQueuedChar:203
20200701 114421.483 (3054580816)       (6)< SocketBuffer_getQueuedChar:222 (-22)
20200701 114421.483 (3054580816)       (6)> SocketBuffer_queueChar:288
20200701 114421.483 (3054580816)       (6)< SocketBuffer_queueChar:313
20200701 114421.483 (3054580816)      (5)< Socket_getch:345 (0)
20200701 114421.483 (3054580816)     (4)< WebSocket_getch:513 (0)
20200701 114421.483 (3054580816)    (3)< MQTTPacket_decode:326 (0)
20200701 114421.483 (3054580816)    (3)> WebSocket_getdata:533
20200701 114421.485 (3054580816)     (4)> Socket_getdata:363
20200701 114421.485 (3054580816)      (5)> SocketBuffer_getQueuedData:163
20200701 114421.485 (3054580816)      (5)< SocketBuffer_getQueuedData:188
20200701 114421.485 (3054580816)      (5)> SocketBuffer_complete:263
20200701 114421.485 (3054580816)      (5)< SocketBuffer_complete:273
20200701 114421.485 (3054580816)     (4)< Socket_getdata:397
20200701 114421.485 (3054580816)    (3)< WebSocket_getdata:595 (1)
20200701 114421.485 (3054580816)    (3)> MQTTPacket_connack:145
20200701 114421.485 (3054580816)    (3)< MQTTPacket_connack:175
20200701 114421.485 (3054580816)   (2)< MQTTPacket_Factory:164 (0)
20200701 114421.485 (3054580816)   (2)> MQTTAsync_retry:3575
20200701 114421.485 (3054580816)    (3)> MQTTProtocol_retry:730
20200701 114421.485 (3054580816)    (3)< MQTTProtocol_retry:749
20200701 114421.485 (3054580816)   (2)< MQTTAsync_retry:3585
20200701 114421.485 (3054580816)  (1)< MQTTAsync_cycle:3893 (0)
20200701 114421.485 (3054580816)  (1)> MQTTAsync_completeConnection:2090
20200701 114421.485 4 async_publish <- CONNACK rc: 0
20200701 114421.485 (3054580816)   (2)> MQTTAsync_cleanSession:2690
20200701 114421.485 (3054580816)    (3)> MQTTPersistence_clear:160
20200701 114421.485 (3054580816)    (3)< MQTTPersistence_clear:164 (0)
20200701 114421.487 (3054580816)    (3)> MQTTProtocol_emptyMessageList:812
20200701 114421.487 (3054580816)    (3)< MQTTProtocol_emptyMessageList:821
20200701 114421.487 (3054580816)    (3)> MQTTProtocol_emptyMessageList:812
20200701 114421.487 (3054580816)    (3)< MQTTProtocol_emptyMessageList:821
20200701 114421.487 (3054580816)    (3)> MQTTAsync_emptyMessageQueue:1894
20200701 114421.487 (3054580816)    (3)< MQTTAsync_emptyMessageQueue:1908
20200701 114421.487 (3054580816)    (3)> MQTTAsync_removeResponsesAndCommands:1918
20200701 114421.487 0 responses removed for client async_publish
20200701 114421.487 0 commands removed for client async_publish
20200701 114421.487 (3054580816)    (3)< MQTTAsync_removeResponsesAndCommands:2003
20200701 114421.487 (3054580816)   (2)< MQTTAsync_cleanSession:2706 (0)
20200701 114421.487 (3054580816)   (2)> Thread_signal_cond:395
20200701 114421.487 (3054580816)   (2)< Thread_signal_cond:400 (0)
20200701 114421.487 (3054580816)  (1)< MQTTAsync_completeConnection:2128 (0)
20200701 114421.487 Calling connect success for client async_publish
20200701 114421.487 (3062973520)  (1)< Thread_wait_cond:421 (0)
20200701 114421.487 (3062973520)  (1)> MQTTAsync_checkTimeouts:1804
20200701 114421.487 [cpp] on_success
20200701 114421.487 (3062973520)  (1)< MQTTAsync_checkTimeouts:1852
20200701 114421.487 (3062973520)  (1)> Thread_wait_cond:413
20200701 114421.489 Calling connected for client async_publish
20200701 114421.489 [cpp] on_connected
20200701 114421.489 (3054580816)  (1)> MQTTPacket_freeConnack:186
20200701 114421.489 (3054580816)  (1)< MQTTPacket_freeConnack:190
20200701 114421.489 (3054580816)  (1)> MQTTAsync_cycle:3742
20200701 114421.489 (3054580816)   (2)> Socket_getReadySocket:237
20200701 114421.489 (3069164720) (0)> MQTTAsync_isConnected:3201
20200701 114421.489 (3069164720) (0)< MQTTAsync_isConnected:3206 (1)
20200701 114421.489 (3069164720) (0)> MQTTAsync_isConnected:3201
20200701 114421.489 (3069164720) (0)< MQTTAsync_isConnected:3206 (1)

<snip>

20200701 114624.669 (3069164720) (0)> MQTTAsync_isConnected:3201
20200701 114624.669 (3069164720) (0)< MQTTAsync_isConnected:3206 (0)
20200701 114624.669 (3062973520)  (1)< Thread_wait_cond:421 (110)
20200701 114624.669 (3062973520)  (1)> MQTTAsync_checkTimeouts:1804
20200701 114624.669 Automatically attempting to reconnect
20200701 114624.669 (3062973520)   (2)> MQTTAsync_addCommand:1050
20200701 114624.669 (3062973520)    (3)> Thread_signal_cond:395
20200701 114624.669 (3062973520)    (3)< Thread_signal_cond:400 (0)
20200701 114624.669 (3062973520)   (2)< MQTTAsync_addCommand:1084 (0)
20200701 114624.669 (3062973520)  (1)< MQTTAsync_checkTimeouts:1852
20200701 114624.669 (3062973520)  (1)> MQTTAsync_processCommand:1398
20200701 114624.669 Connecting to serverURI test.mosquitto.org:1883 with MQTT version 4
20200701 114624.669 (3062973520)   (2)> MQTTProtocol_connect:114
20200701 114624.669 (3062973520)    (3)> MQTTProtocol_addressPort:58
20200701 114624.669 (3062973520)    (3)< MQTTProtocol_addressPort:90
20200701 114624.669 (3062973520)    (3)> Socket_new:668
20200701 114624.769 (3054580816)   (2)< MQTTAsync_sleep:446
20200701 114624.769 (3069164720) (0)> MQTTAsync_isConnected:3201
20200701 114624.769 getaddrinfo failed for addr test.mosquitto.org with rc -3
20200701 114624.769 test.mosquitto.org is not a valid IP address
20200701 114624.769 (3062973520)    (3)< Socket_new:785 (-3)
20200701 114624.769 (3062973520)   (2)< MQTTProtocol_connect:156 (-3)
20200701 114624.769 (3062973520)   (2)> MQTTAsync_disconnect1:3130
20200701 114624.769 (3062973520)   (2)< MQTTAsync_disconnect1:3167 (-3)
20200701 114624.769 (3062973520)   (2)> MQTTAsync_checkConn:526
20200701 114624.769 (3062973520)   (2)< MQTTAsync_checkConn:529 (1)
20200701 114624.769 Connect failed, more to try
20200701 114624.769 (3062973520)   (2)> MQTTAsync_addCommand:1050
20200701 114624.769 (3062973520)    (3)> Thread_signal_cond:395
20200701 114624.769 (3062973520)    (3)< Thread_signal_cond:400 (0)
20200701 114624.769 (3062973520)   (2)< MQTTAsync_addCommand:1084 (0)
20200701 114624.769 (3062973520)  (1)< MQTTAsync_processCommand:1734 (1)
20200701 114624.769 (3062973520)  (1)> MQTTAsync_processCommand:1398
20200701 114624.769 (3054580816)   (2)> MQTTAsync_retry:3575
20200701 114624.769 (3054580816)    (3)> MQTTProtocol_keepalive:608
20200701 114624.769 (3054580816)    (3)< MQTTProtocol_keepalive:644
20200701 114644.789 (3054580816)    (3)> MQTTProtocol_retry:730
20200701 114644.789 (3054580816)    (3)< MQTTProtocol_retry:749
20200701 114644.789 (3054580816)   (2)< MQTTAsync_retry:3585
20200701 114644.789 (3054580816)  (1)< MQTTAsync_cycle:3893 (-1)
20200701 114644.789 (3054580816)  (1)> MQTTAsync_cycle:3742
20200701 114644.789 (3069164720) (0)< MQTTAsync_isConnected:3206 (0)
20200701 114644.789 Connecting to serverURI test.mosquitto.org:1883 with MQTT version 3
20200701 114644.789 (3069164720) (0)> MQTTAsync_isConnected:3201
20200701 114644.789 (3062973520)   (2)> MQTTProtocol_connect:114
20200701 114644.789 (3062973520)    (3)> MQTTProtocol_addressPort:58
20200701 114644.789 (3062973520)    (3)< MQTTProtocol_addressPort:90
20200701 114644.789 (3062973520)    (3)> Socket_new:668
20200701 114644.789 (3054580816)   (2)> Socket_getReadySocket:237
20200701 114644.789 (3054580816)   (2)< Socket_getReadySocket:309 (0)
20200701 114644.789 (3054580816)   (2)> MQTTAsync_sleep:440
20200701 114644.789 (3054580816)   (2)< MQTTAsync_sleep:446
20200701 114644.789 getaddrinfo failed for addr test.mosquitto.org with rc -3
20200701 114644.789 test.mosquitto.org is not a valid IP address
20200701 114644.789 (3062973520)    (3)< Socket_new:785 (-3)
20200701 114644.789 (3062973520)   (2)< MQTTProtocol_connect:156 (-3)
20200701 114704.812 (3062973520)   (2)> MQTTAsync_disconnect1:3130
20200701 114704.812 (3062973520)   (2)< MQTTAsync_disconnect1:3167 (-3)
20200701 114704.812 (3062973520)   (2)> MQTTAsync_checkConn:526
20200701 114704.812 (3062973520)   (2)< MQTTAsync_checkConn:529 (0)
20200701 114704.812 Calling command failure for client async_publish
20200701 114704.812 [cpp] on_failure
20200701 114704.812 (3062973520)   (2)> MQTTProperties_free:366
20200701 114704.812 (3062973520)   (2)< MQTTProperties_free:389
20200701 114704.812 (3062973520)  (1)< MQTTAsync_processCommand:1734 (1)
20200701 114704.812 (3062973520)  (1)> Thread_wait_cond:413
20200701 114704.812 (3054580816)   (2)> MQTTAsync_retry:3575
20200701 114704.812 (3054580816)    (3)> MQTTProtocol_keepalive:608
20200701 114704.812 (3054580816)    (3)< MQTTProtocol_keepalive:644
20200701 114704.812 (3054580816)    (3)> MQTTProtocol_retry:730
20200701 114704.812 (3054580816)    (3)< MQTTProtocol_retry:749
20200701 114704.812 (3054580816)   (2)< MQTTAsync_retry:3585
20200701 114704.812 (3054580816)  (1)< MQTTAsync_cycle:3893 (-1)
20200701 114704.812 (3054580816)  (1)> MQTTAsync_cycle:3742
20200701 114704.812 (3054580816)   (2)> Socket_getReadySocket:237
20200701 114704.812 (3054580816)   (2)< Socket_getReadySocket:309 (0)
20200701 114704.814 (3054580816)   (2)> MQTTAsync_sleep:440
20200701 114704.814 (3069164720) (0)< MQTTAsync_isConnected:3206 (0)
20200701 114704.814 (3054580816)   (2)< MQTTAsync_sleep:446
20200701 114704.814 (3054580816)   (2)> MQTTAsync_retry:3575
20200701 114704.814 (3054580816)    (3)> MQTTProtocol_retry:730
20200701 114704.814 (3054580816)    (3)< MQTTProtocol_retry:749
20200701 114704.814 (3054580816)   (2)< MQTTAsync_retry:3585
20200701 114704.814 (3054580816)  (1)< MQTTAsync_cycle:3893 (-1)
20200701 114704.814 (3054580816)  (1)> MQTTAsync_cycle:3742
20200701 114704.814 (3054580816)   (2)> Socket_getReadySocket:237
20200701 114704.814 (3054580816)   (2)< Socket_getReadySocket:309 (0)
20200701 114704.814 (3054580816)   (2)> MQTTAsync_sleep:440
20200701 114704.814 (3069164720) (0)> MQTTAsync_isConnected:3201
20200701 114704.814 (3069164720) (0)< MQTTAsync_isConnected:3206 (0)

<snip>
```

**Environment**

* OS: A custom Linux build for Phoenix Contact's AXC F 2152 controller.
* Version: 2020.3.1

**Additional context**

I suspect that the problem may be related to issue https://github.com/eclipse/paho.mqtt.c/issues/690, but I could be wrong.

**Comment by icraggs** [comment]
I think the call is taking time because you pulled the cable causing the DNS lookup to fail but only after a timeout.  Cross platform asynchronous DNS resolution is not straightforward which is why is hasn't been implemented so far.  Most of the time, people are dealing with TCP connection breaks or MQTT server downtime, which is why it hasn't risen to be a crucial issue.

if that is the issue, then possible solutions at the moment are:

- avoid DNS lookups by using the IP address, or adding the IP address to the hosts table
- use the connected() callback to determine when the re-connection has succeeded rather than calling isconnected()

If it's just the isconnected() call that's affected, then there may be a simpler solution for that.

**Comment by martinboers** [comment]
OK, thanks. It's not only the is_connected call, but I have confirmed that this (and all my other problems around disconnect) are probably due to this DNS lookup issue, which - as you suggest - can be avoided by using the IP address.

Thanks for your quick response and very helpful answer.

**Comment by icraggs** [comment]
Reopening and changing the title as a reminder for me or someone to look into the question of async DNS lookups at some point.

**Comment by jumoog** [comment]
maybe using https://github.com/c-ares/c-ares ?

**Comment by icraggs** [comment]
@jumoog yes I'd thought about using that in the past.  I thought there was an obstacle but I can't remember what it was.

**Comment by keysight-daryl** [comment]
other option is to just thread the dns call internally and cache the result as rarely the ip would change if it ever gets a valid result. other option is for the caller to just thread their 'status loop' code

**Comment by icraggs** [comment]
Would also fix #690

**Comment by icraggs** [comment]
I'm going to trackk this under #690.
